### PR TITLE
Add delete room admin endpoint

### PR DESCRIPTION
**Pull request by dklimpel**
The Delete Room admin API allows server admins to remove rooms from server
and block these rooms.
`DELETE /_synapse/admin/v1/rooms/<room_id>`
It is a combination and improvement of "[Shutdown room](https://github.com/matrix-org/synapse/blob/develop/docs/admin_api/shutdown_room.md)" and "[Purge room](https://github.com/matrix-org/synapse/blob/develop/docs/admin_api/purge_room.md)" API.

Fixes: #6425 

It also fix a bug in [synapse/storage/data_stores/main/room.py](synapse/storage/data_stores/main/room.py) in ` get_room_with_stats`.
It should return `None` if the room is unknown. But it returns an `IndexError`.
https://github.com/matrix-org/synapse/blob/901b1fa561e3cc661d78aa96d59802cf2078cb0d/synapse/storage/data_stores/main/room.py#L99-L105

Related to:
- #5575
- https://github.com/Awesome-Technologies/synapse-admin/issues/17

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

**Comment by dklimpel** [comment]
> rather than cutting-and-pasting the implementation of ShutdownRoom, please could you factor out the code to a `handler` so that it can be shared?
> 
> By the way, please could you join `#synapse-dev:matrix.org` so that we can collaborate more easily on some of this stuff? It's not very high-volume.

I have moved it to an own handler.
The room handler has not tests at the moment.

File: docs/admin_api/rooms.md [hunk 1]
```diff
@@ -318,3 +318,89 @@ Response:
...
+The following parameters should be set in the URL:
+
+* `room_id` - The ID of the room.
+
+The following JSON body parameters are available:
+
+* `new_room_user_id` - Optional. A string representing the user ID of the user that will admin
+                       the new room that all users in the old room will be moved to. If not
+                       set the users will not be moved to a new room and only leave the old room
+                       without any information. Defaults to `None`.
```
Comment (richvdh) [inline docs/admin_api/rooms.md:358]:
```suggestion
* `new_room_user_id` - Optional. If set, a new room will be created with this user ID
      as the creator and admin, and all users in the old room will be moved into that
      room. If not set, no new room will be created and the users will just be removed
      from the old room.
```

File: docs/admin_api/rooms.md [hunk 2]
```diff
@@ -318,3 +318,89 @@ Response:
...
+* `room_name` - Optional. A string representing the name of the room that new users will be
+                invited to. Defaults to `Content Violation Notification`
+* `message` - Optional. A string containing the first message that will be sent as
+              `new_room_user_id` in the new room. Ideally this will clearly convey why the
+               original room was shut down. Defaults to `Sharing illegal content on this server
+               is not permitted and rooms in violation will be blocked.`
+* `block` - Optional. A boolean if `room_id` will be set on blocking list. The room will be
+            blocked for this server and preventing new joins. Defaults to `True`.
```
Comment (richvdh) [inline docs/admin_api/rooms.md:366]:
```suggestion
* `block` - Optional. If set to `true`, this room will be added to a blocking list, preventing future attempts to
  join the room. Defaults to `true`.
```

File: synapse/rest/admin/rooms.py [hunk 3]
```diff
@@ -46,137 +47,77 @@ class ShutdownRoomRestServlet(RestServlet):
...
-                await self.room_member_handler.update_membership(
-                    requester=target_requester,
-                    target=target_requester.user,
-                    room_id=new_room_id,
-                    action=Membership.JOIN,
-                    content={},
-                    ratelimit=False,
-                    require_consent=False,
-                )
+        content = parse_json_object_from_request(request, allow_empty_body=True)
```
Comment (richvdh) [inline synapse/rest/admin/rooms.py:97]: I don't think there is any need to allow an empty body here?

File: docs/admin_api/rooms.md [hunk 4]
```diff
@@ -318,3 +318,89 @@ Response:
...
+
+The following fields are possible in the JSON response body:
```
Comment (richvdh) [inline docs/admin_api/rooms.md:368]:
```suggestion
The following fields are returned in the JSON response body:
```

File: docs/admin_api/rooms.md (hunk 2 above)
Comment (richvdh) [inline docs/admin_api/rooms.md:366]: wouldn't it make more sense for this to default to false?

File: synapse/handlers/room.py [hunk 5]
```diff
@@ -1091,3 +1093,168 @@ def get_current_key(self):
...
+        if not new_room_name:
+            new_room_name = self.DEFAULT_ROOM_NAME
+        if not message:
+            message = self.DEFAULT_MESSAGE
+
+        # Check RoomID
+        if not RoomID.is_valid(room_id):
+            raise SynapseError(400, "%s was not legal room ID" % (room_id,))
+
+        if not await self.store.get_room_with_stats(room_id):
```
[the remark this answers is not in the document]
Comment (dklimpel) [reply synapse/handlers/room.py:1134]: This serves to check that the room exists on the server and has data.

File: synapse/handlers/room.py [hunk 6]
```diff
@@ -1091,3 +1093,168 @@ def get_current_key(self):
...
+        if block:
+            await self.store.block_room(room_id, requester_user_id)
+
+        if new_room_user_id:
+            if not self.hs.is_mine_id(new_room_user_id):
+                raise SynapseError(
+                    400, "User must be our own: %s" % (new_room_user_id,)
+                )
+
+            if not await self.admin_handler.get_user(
```
[the remark this answers is not in the document]
Comment (dklimpel) [reply synapse/handlers/room.py:1148]: I can change it to `store.get_user_by_id`. It is necessary that a valid user creates the new room.

File: synapse/handlers/room.py [hunk 7]
```diff
@@ -1091,3 +1093,168 @@ def get_current_key(self):
...
+
+        if not await self.store.get_room_with_stats(room_id):
+            raise NotFoundError("Unknown room id %s" % (room_id,))
+
+        # This will work even if the room is already blocked, but that is
+        # desirable in case the first attempt at blocking the room failed below.
+        if block:
+            await self.store.block_room(room_id, requester_user_id)
+
+        if new_room_user_id:
```
[the remark this answers is not in the document]
Comment (dklimpel) [reply synapse/handlers/room.py:1142]: The reason for the new API is to allow a room to be removed from a server. This does not necessarily mean moving all users to a new room. The admin has the choice to move the users or not. In `shutdown_room` API is `new_room_user_id` required.

File: synapse/rest/admin/rooms.py (hunk 3 above)
Comment (dklimpel) [reply synapse/rest/admin/rooms.py:97]: The empty body is needed when an admin want to delete the room from server and do not want to move the users into a new room.

File: docs/admin_api/rooms.md [hunk 8]
```diff
@@ -318,3 +318,89 @@ Response:
...
+{
+    "new_room_user_id": "@someuser:example.com",
+    "room_name": "Content Violation Notification",
+    "message": "Bad Room has been shutdown due to content violations on this server. Please review our Terms of Service.",
+    "block": true
+}
+```
+
+To use it, you will need to authenticate by providing an ``access_token`` for a
+server admin: see `README.rst <README.rst>`_.
```
[the remark this answers is not in the document]
Comment (dklimpel) [reply docs/admin_api/rooms.md]: I have moved it up.

File: docs/admin_api/rooms.md [hunk 9]
```diff
@@ -318,3 +318,89 @@ Response:
...
+
+A response body like the following is returned:
```
[the remark this answers is not in the document]
Comment (dklimpel) [reply docs/admin_api/rooms.md:366]: Parameters are at the end now and we do not need a new heading.

File: synapse/handlers/room.py [hunk 10]
```diff
@@ -1091,3 +1093,168 @@ def get_current_key(self):
...
+                )
+
+                await self.room_member_handler.forget(target_requester.user, room_id)
+
+                kicked_users.append(user_id)
+            except Exception:
+                logger.exception("Failed to leave old room for %r", user_id)
+                failed_to_kick_users.append(user_id)
+
+            # Join users to new room
```
[the remark this answers is not in the document]
Comment (dklimpel) [reply synapse/handlers/room.py]: You are right. I have changed it.

**Comment by dklimpel** [comment]
@richvdh: ready for review

File: synapse/rest/admin/rooms.py (hunk 3 above)
Comment (richvdh) [reply synapse/rest/admin/rooms.py:96]: why can the admin not set the body to `{}` in that case?

File: synapse/rest/admin/rooms.py [hunk 11]
```diff
@@ -46,137 +46,76 @@ class ShutdownRoomRestServlet(RestServlet):
...
-        )
-        new_room_id = info["room_id"]
-
-        requester_user_id = requester.user.to_string()
 
-        logger.info(
-            "Shutting down room %r, joining to new room: %r", room_id, new_room_id
+        ret = await self.room_shutdown_handler.shutdown_room(
+            room_id=room_id,
+            new_room_user_id=content.get("new_room_user_id"),
```
Comment (richvdh) [inline synapse/rest/admin/rooms.py:63]:
```suggestion
            new_room_user_id=content["new_room_user_id"],
```

File: synapse/handlers/room.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/handlers/room.py:1184]: >  In shutdown_room API is new_room_user_id required.

we don't appear to check that it is set to a non-empty string though. I suggest:

```suggestion
        if new_room_user_id is not None:
```

File: synapse/handlers/room.py [hunk 12]
```diff
@@ -1089,3 +1091,208 @@ def get_current_key(self):
...
+            await self.store.block_room(room_id, requester_user_id)
+
+        if new_room_user_id:
+            if not self.hs.is_mine_id(new_room_user_id):
+                raise SynapseError(
+                    400, "User must be our own: %s" % (new_room_user_id,)
+                )
+
+            if not await self.store.get_user_by_id(new_room_user_id):
+                raise NotFoundError("Unknown user %s" % (new_room_user_id,))
```
Comment (richvdh) [inline synapse/handlers/room.py:1191]: I'm not keen on this change to the behaviour of `shutdown_room`. Currently it is not necessary for the user to be registered.

Comment (dklimpel) [reply synapse/handlers/room.py]: I did some tests, changed it and adapted the documentation.

File: synapse/rest/admin/rooms.py (hunk 3 above)
Comment (dklimpel) [reply synapse/rest/admin/rooms.py]: I have changed it and added a hint to the documentation.

**Comment by ell1e** [comment]
Just a feedback from reading this pull request as an outsider:

>  It is a combination and improvement of "Shutdown room" and "Purge room" API.

The "improvement" that also made it into the documentation description of this new function seems quite confusing to me. To me it looks like essentially this new "delete" just does shutdown followed by purge, is that not all it does?  If the "improvement" just means the code is cleaned up as well then I suggest it should be removed from the code comment / documentation description since it seems irrelevant and misleading at best in there. Otherwise I suggest that it is clarified what exactly this "improvement" means in terms of actual behavior.

File: docs/admin_api/rooms.md [hunk 13]
```diff
@@ -390,7 +390,7 @@ The following JSON body parameters are available:
 * `new_room_user_id` - Optional. If set, a new room will be created with this user ID
       as the creator and admin, and all users in the old room will be moved into that
       room. If not set, no new room will be created and the users will just be removed
-      from the old room.
+      from the old room. The user ID must be from local server but it does not have to exist.
```
Comment (richvdh) [inline docs/admin_api/rooms.md:393]: this is confusingly worded.

```suggestion
      from the old room. The user ID must be on the local server, but does not necessarily
      have to belong to a registered user.
```

**Comment by richvdh** [comment]
@ell1e : I don't disagree: that line isn't terribly clear. That said, I don't want to hold this PR up on another round of documentation tweaking. 

For the record, the new API is slightly more flexible than using the two existing APIs separately (for example: `Shutdown room` automatically moves all existing users to a new room. That is optional with the new APIs.) There is, of course, also a good argument that these improvements should be ported back to the shutdown room api...

(Aside: when you comment on a PR,  please can you comment on a line in the "files changed" tab.  That (a) helps people see which bit of the PR you're talking about, and (b) means that the conversation is threaded rather than being mixed in with all the other traffic on the PR.)

**Comment by richvdh** [comment]
Ah bother. @dklimpel I don't have sign-off from you. Please can you add a `signed-off-by` comment here as per https://github.com/matrix-org/synapse/blob/develop/CONTRIBUTING.md#sign-off?

**Comment by dklimpel** [comment]
> Ah bother. @dklimpel I don't have sign-off from you. Please can you add a `signed-off-by` comment here as per https://github.com/matrix-org/synapse/blob/develop/CONTRIBUTING.md#sign-off?

@richvdh I have added it https://github.com/matrix-org/synapse/pull/7613#issue-426050177

**Pull request merged**